NEP50

ev-br opened the pull request:
Split off https://github.com/cupy/cupy/pull/8314 to keep the diff sizes under control.

Adapt the codebase to NEP 50 (https://numpy.org/neps/nep-0050-scalar-promotion.html). Unlike gh-8314, this PR can be run under NumPy 1.x using 

```
$ export NPY_PROMOTION_STATE=weak
$ pytest ...
```

The main change is that under NEP 50, we have to keep track of "weak" scalars:

```
In [5]: (np.int8(1) + np.uint8(1)).dtype
Out[5]: dtype('int16')

In [6]: (1 + np.uint8(1)).dtype
Out[6]: dtype('uint8')
```

Implementation-wise, CuPy seems to convert python scalars to numpy scalars rather early in the game; thus we keep track of which arguments of ufuncs are python scalars, and take this into account when selecting ufunc loops. 
So far am only doing it in ufuncs, will need to see if ElementwiseKernels / fusion loops / reductions are affected. 

The goal is to make tests pass locally first with numpy 1.26.x & weak promotion rules, so that one can think of switching this on the CI. So far it's not even close: need to weed away lots of assumptions about the promotion rules in code and tests.

leofang commented:
cc: @seberg for vis

kmaehashi commented:
Thank you for tackling this, @ev-br! As for OOMs, I'd suggest excluding slow-labeled tests (`pytest -m 'not slow'`) during development. They are for testing functionality against huge ndarray and I think they are safely ignored when working on type promotional rules.

seberg commented on cupy/linalg/_einsum.py:
```diff
@@ -150,6 +150,9 @@ def _parse_einsum_input(args):
                 'specified in the subscripts string')
 
+        # NumPy ignores 'weak' scalars and always returns i64/f64
+        operands = [cupy.asarray(op) for op in operands]
```
Maybe should change that one day, but not today.  `i64` I assume is true, `f64` I would think not if inputs are float32?

I suppose since scalars are an odd input to einsum, not changing it should be OK or maybe we should fix this one in NumPy?

ev-br replied:
Indeed, `s/always/sometimes`:

```
In [5]: np.einsum('i->', np.arange(8, dtype=np.float32)).dtype
Out[5]: dtype('float32')

In [6]: np.einsum(', i->', 3, np.arange(8, dtype=np.float32)).dtype
Out[6]: dtype('float64')

In [7]: np.__version__
Out[7]: '2.0.0rc2'
```

numpy bug?

ev-br replied:
So this patch correctly avoids using weak promotion in einsum.

seberg commented on tests/cupy_tests/random_tests/test_sample.py:
```diff
@@ -114,7 +114,7 @@ class TestRandintDtype(unittest.TestCase):
         size = (1000,)
         low = numpy.iinfo(dtype).min
-        high = numpy.iinfo(dtype).max + 1
+        high = numpy.iinfo(dtype).max
```
Hmmm, this change the test, Maybe need at least a new xfail here since that should work?

ev-br replied:
Indeed. Have to xfail for numpy 1.26 + remove short ufunc loops here as well:
https://github.com/cupy/cupy/pull/8323/commits/7217c8dc2ff9269c4aeffaa31945ef8c13b8705c

seberg commented on cupy/_core/_kernel.pyx:
```diff
@@ -1584,21 +1613,18 @@ cdef class _Ops:
+                        for a in in_args])
+            in_types = tuple([a.dtype.type for a in in_args])
+
+            op = cache.get((in_types, weaks), ())
```
Makes sense, I suspect one could use `in_types[weak_pos] = int` or so also here, but I think using the `weak` tuple works well for starters.
(I might want to push for a cleanup after/with my other PR, but this is more urgent.)

seberg commented on cupy/_core/_kernel.pyx:
```diff
@@ -1584,21 +1613,18 @@ cdef class _Ops:
+            _Ops out_ops):
         cdef _Ops ops_
         if dtype is None:
-            use_raw_value = _check_should_use_min_scalar(in_args)
```
Not for me to decide.  If we wish to support both 1.x and 2.x (for the most part), it could make sense to keep the `_check_should_use_min_scalars` here.

As a very subtle point, NumPy still (usually) uses the same check as in `if all(weak): weak = [False] * len(weak)`.

(Which is arguably annoying, and only has an effect for weird integer inputs, though, so I think it is OK to ignore it.)

ev-br replied:
Not for me to decide either. This PR so far labored under assumption that cupy 14 only supports numpy 2.0, per https://github.com/cupy/cupy/issues/8306#issuecomment-2076426835.
BTW, this implies that cupyx.scipy requires scipy 1.13

seberg commented on cupy/_core/_kernel.pyx:
```diff
@@ -1610,6 +1636,35 @@ cdef class _Ops:
+
+                        if not ok:
+                            raise OverflowError(f"Python integer {arg} out of "
+                                                f"bounds for {out_type}.")
```
This looks like a lot of code that really ends up doing the same as: `out_type(int(arg))` or `out_type(arg.item())`?
I suppose one could check safe-cast first, although I am not sure it would be faster.

ev-br replied:
Not sure what you mean, mind expanding? This code is meant to check `iinfo.min <= arg <= iinfo.max` where arg can be signed or unsigned. Cython is trying to be helpful, so need explicit casts to avoid overflows.

seberg replied:
What I mean is, write this whole thing as:
```
for out_type, arg, weak for zip(out_types, in_args, weaks):
    if not weak is not int:  # assume we change what is stored in `weaks`
        continue

    out_type(int(in_arg))  # check if Python integer fits
```

seberg replied:
now that I look again, I think it also needs to loop `in_types` and not rely on `out_type` being identical.

seberg commented on cupy/_core/_kernel.pyx:
```diff
@@ -1618,20 +1673,49 @@ cdef class _Ops:
                     if not can_cast(it[0], ot) and not can_cast(it[1], ot):
                         break
-                elif not can_cast(it, ot):
+                elif not is_weak and not can_cast(it, ot):
```
I know it is a bit slow, but it should be cached, so I would suggest replacing the full above logic logic with:
```
try:
    safe = np.result_type(weak(0), ot) == ot
except TypeError:
    safe = False
```
Might not be the prettiest ever, but I think it is preferable to custom implementation of the category logic and more flexible for future additions.

I am suggesting another trick:  Let's make `weaks = [False, int, float, complex]` i.e. store either `False` _or_ the type, so that we can use the type information directly from there?

ev-br replied:
Currently, `in_types` entries are numpy scalars types already, after the call to https://github.com/cupy/cupy/blob/622cdf4ae126d8bf611eeb0e0dd77a5e069b5cd5/cupy/_core/_kernel.pyx#L180 at https://github.com/cupy/cupy/blob/622cdf4ae126d8bf611eeb0e0dd77a5e069b5cd5/cupy/_core/_kernel.pyx#L1313

Which is the main reason to have an additional `weaks` tuple to be able to distinguish `int8(1) + 1` from `int8(1) + int8(1)`.

So your suggestion is to keep python types to be python types all the way instead of converting them to numpy types early on?

seberg commented on cupy/_core/_routines_math.pyx:
```diff
@@ -1037,11 +1037,15 @@ _subtract = create_arithmetic(
-    ('bb->d', 'BB->d', 'hh->d', 'HH->d', 'ii->d', 'II->d', 'll->d', 'LL->d',
-     'qq->d', 'QQ->d', 'ee->e', 'ff->f', 'dd->d', 'FF->F', 'DD->D'),
-    'out0 = (out0_type)in0 / (out0_type)in1',
+    ('qq->d', 'QQ->d',
```
Right, this has a slight miss for mixed signs, but that seems OK.

In principle it would be nice to solve this with a custom promotion step for the ufunc.  This could be very nice, because for (almost!) all NumPy ufunc, we should be able to at least try and use:
```
dtypes = ufunc.resolve_dtypes()
```

seberg reviewed (commented):
Thanks so much for starting this @ev-br!  Sorry about the slower followup on our discussions.
This is nicer and really quite minimal, and I think we can make it even more minimal!

As said below, I think passing the `weak` tuple is probably the pragmatic solution right now and may well be the best solution in the long run also. 

A few notes (some of which are inline):
* It might be nice to have at least a single per-ufunc/kernel "promotion" step.  I.e. a custom `guess_loop`.  That would solve many issues, because you could call `np.ufunc.resolve_dtypes()`.  I will note that this function *is* currently slow and also is only available since 1.24 or so.
* I have suggested some simplifications, which I *think* should make things even shorter (I may be missing reasons for why not though).
* In NumPy I had to add some "very special" paths unfortunately.  I think it makes sense to skip that for this PR.  However, it would probably make sense to at least for `uint_array > -1` add a special path to make it work.
  (i.e. if it is tricky to do for the comparison ufuncs, make it work for the comparisons.  For NumPy downstream testing code ran into it often enough and it is somewhat surprising to fail...)

ev-br commented:
Progress: `@fuse` reuses the ufunc machinery for type promotion. `core_tests/fusion_tests` all pass, at least locally.

ev-br commented:
Status:

-  locally,``$ for f in `ls tests/cupy_tests; do echo $f && pytest tests/cupy_tests/$f -n 8 -m'not slow' --tb=no; done`` all passes modulo two OOMs in `test_numpy_interop`. OOMs are likely irrelevant for the promotion state changes;
- @seberg's comments are all addressed --- mostly by Sebastian himself, his commits are included into the PR;
- there are several skips/xfails for things which are broken in numpy 1.26.x + weak promotion and are fixed in numpy 2.0.0rc2;
- there might be more issues in the `cupyx` namespace, but then it also depends on e.g. the scipy version;
- there's a question on how to enable this, also in relation to gh-8314:
     - do we want to push this PR to be CI green (then how to add `export NPY_PROMOTION_STATE=weak` on CI?) , then work on NumPy 2.0 API changes separately, or
     - can we tolerate some CI failures while this is in flux, or
     - do we lump these two PRs together.

seberg commented on cupy/_core/fusion.pyx:
```diff
@@ -597,34 +558,40 @@ class _FusionHistory(object):
-        # TODO(asi1024): Fix to use ``guess_routine``.
-        for op in ufunc._ops.ops:
+        in_sctypes = tuple([a.dtype.type for a in in_vars])
+        weaks = tuple([a.is_weak if hasattr(a, 'is_weak') else False
+                       for a in in_vars])
```
```suggestion
        weaks = tuple([getattr(a, 'is_weak', False) for a in in_vars])
```

seberg commented on tests/cupyx_tests/jit_tests/test_cooperative_groups.py:
```diff
@@ -78,12 +78,15 @@ def test_grid(x):
         assert x[1] == 64
         assert (x[2], x[3], x[4]) == (2, 1, 1)
-        assert (x[5:] == 2**64-1).all()
+        # TODO (ev-br): revert back to x[5:] == -1 after the edge case of
```
```suggestion
        # XXX: np2.0: revert back to x[5:] == -1 after the edge case of
```
Let's make sure this is grep'able

seberg reviewed (commented):
From my perspective, I think the logic here should be big enough step forward that it would be great to push it forward/merge soon.

The main issue (as @ev-br said) seems now to ensure that tests pass for developers and CI.  What we need for that may be:
* Force testing with NumPy 1.26 and `NPY_PROMOTION_STATE="weak"`.  We can do that in `conftest.py`.
* Fixing CI!  Unfortunately, I am not sure what is all needed for that.  I would think that could be split out in a separate PR.

Otherwise, I think it should be at most be small fixups (a few new xfails, and maybe making sure all xfails/odd or temporary paths are easy to grep for).

There are some more complicated issues, however, I think it is probably better to defer them:
* Many current xfails actually work in NumPy 2 (should we use a skipif?)
* NumPy 2 has some special logic to e.g. make `uint_arr > -1` work.
  (I am not sure how vital this is, but to fix it we will have to add a distinct layer of logic and that seems better in a new PR with an issue.)

---

The goal here is currently to create a CuPy v14 that matches NumPy 2, but fails to match NumPy 1.x w.r.t. to promotion.
That means users will have to stay with CuPy v13 if they have NumPy 1.x (and need it to match).
I think this aligns with the current plans of the CuPy team, but I am not sure.

I will note that I am sure we could implement *both* versions side-by-side without too much hassle and I would be happy to do this if it seems helpful/desirable (even if it is just helpful with getting CI going).

ev-br commented on install/cupy_builder/_features.py:
```diff
@@ -422,7 +422,7 @@ def get_features(ctx: Context) -> Dict[str, Feature]:
             CUDA_cuda(ctx),
             _from_dict(CUDA_cusolver, ctx),
-            _from_dict(CUDA_cudnn, ctx),
+            # _from_dict(CUDA_cudnn, ctx),
```
Note this is needed for me locally (my system has an incompatible cudnn version), should revert the corresponding commit before merging if this goes in.

ev-br commented:
With a slew of fixups/skips/xfails in the cupyx namespace, this is ready from my side.

jakirkham commented:
/test mini

jakirkham commented:
It might be worthwhile to merge in `main`. A few fixes for NumPy 2 went in recently

ev-br commented:
Rebased on main.

jakirkham commented:
/test mini

ev-br commented:
Updated the PR to add a fix in fusion.pyx from @seberg to hopefully fix some of the CI failures.
The majority of other CI "failures" are in fact XPASSes in cupyx.scipy.sparse on scipy 1.13 --- there has been a bit of activity in scipy.sparse recently, and supporting both scipy 1.13 and scipy <= 1.12 gets a little annoying, esp if the test suite is in the strict mode. 

On that note, it would be actually great to have some feedback as to
- whether the general direction of this PR aligns with the project?
- what is the target range of supported numpy/scipy versions. Some CI runs use fairly ancient numpies, also scipy 1.10 etc

jakirkham commented:
/test mini

asi1024 commented:
@ev-br 
Thanks for your investigation!

> whether the general direction of this PR aligns with the project?

The general direction of this PR looks fine, but there's a concern about potential degradation of the host side performance.
- _kernel.pyx: Performance is a significant concern in this situation. Could you make a evaluation by comparing the execution time with the small elementwise kernel call?
- fusion: Since trace only runs on the initial run, performance isn't a matter of concern in this scenario.

> what is the target range of supported numpy/scipy versions. Some CI runs use fairly ancient numpies, also scipy 1.10 etc

We are planning to drop the support of SciPy<=1.12 in CuPy v14.

ev-br commented:
Here's a small very much not scientific benchmark @asi1024 : just the docstring example of https://docs.cupy.dev/en/stable/reference/generated/cupyx.profiler.benchmark.html, with either two arrays or an array and a scalar. It seems that the performance differences between main and this PR are within errorbars:


```
==== on main:

In [18]: import cupy as cp
    ...: from cupyx.profiler import benchmark
    ...: 
    ...: def f(a, b):
    ...:     return 3 * cp.sin(-a) * b
    ...: 
    ...: for n in [10, 100, 1000, 1000]:
    ...:    a = 0.5 - cp.random.random((n,))
    ...:    b = 42 #cp.random.random((n,))
    ...:    print(benchmark(f, (a, b), n_repeat=1000))
    ...: 
f                   :    CPU:    49.006 us   +/-  1.191 (min:    47.510 / max:    57.828) us     GPU-0:    52.712 us   +/-  1.366 (min:    50.944 / max:    61.760) us
f                   :    CPU:    49.129 us   +/-  1.011 (min:    47.620 / max:    55.701) us     GPU-0:    52.846 us   +/-  1.505 (min:    50.176 / max:    81.824) us
f                   :    CPU:    49.977 us   +/-  1.001 (min:    48.501 / max:    64.172) us     GPU-0:    53.684 us   +/-  1.160 (min:    51.296 / max:    69.312) us
f                   :    CPU:    49.644 us   +/-  1.037 (min:    48.250 / max:    63.950) us     GPU-0:    53.324 us   +/-  1.151 (min:    51.200 / max:    67.680) us

In [19]: import cupy as cp
    ...: from cupyx.profiler import benchmark
    ...: 
    ...: def f(a, b):
    ...:     return 3 * cp.sin(-a) * b
    ...: 
    ...: for n in [10, 100, 1000, 1000]:
    ...:    a = 0.5 - cp.random.random((n,))
    ...:    b = cp.random.random((n,))
    ...:    print(benchmark(f, (a, b), n_repeat=1000))
    ...: 
f                   :    CPU:    58.835 us   +/- 16.967 (min:    35.310 / max:   109.483) us     GPU-0:    63.930 us   +/- 18.142 (min:    37.792 / max:   117.920) us
f                   :    CPU:    37.784 us   +/-  1.030 (min:    36.580 / max:    49.932) us     GPU-0:    41.473 us   +/-  1.157 (min:    39.424 / max:    53.600) us
f                   :    CPU:    38.974 us   +/-  1.241 (min:    37.283 / max:    52.651) us     GPU-0:    42.646 us   +/-  1.407 (min:    39.808 / max:    57.248) us
f                   :    CPU:    38.408 us   +/-  1.047 (min:    36.532 / max:    47.401) us     GPU-0:    42.245 us   +/-  1.186 (min:    39.872 / max:    51.136) us


on branch:

In [1]: import cupy as cp
   ...: from cupyx.profiler import benchmark
   ...: 
   ...: def f(a, b):
   ...:     return 3 * cp.sin(-a) * b
   ...: 
   ...: for n in [10, 100, 1000, 1000]:
   ...:    a = 0.5 - cp.random.random((n,))
   ...:    b = 42 #cp.random.random((n,))
   ...:    print(benchmark(f, (a, b), n_repeat=1000))
   ...: 
f                   :    CPU:    48.949 us   +/-  0.904 (min:    47.570 / max:    56.149) us     GPU-0:    52.535 us   +/-  1.100 (min:    50.080 / max:    59.552) us
f                   :    CPU:    48.677 us   +/-  1.390 (min:    46.792 / max:    78.181) us     GPU-0:    52.313 us   +/-  1.509 (min:    49.440 / max:    81.920) us
f                   :    CPU:    49.150 us   +/-  1.051 (min:    47.840 / max:    64.112) us     GPU-0:    52.745 us   +/-  1.204 (min:    51.168 / max:    67.904) us
f                   :    CPU:    49.279 us   +/-  0.853 (min:    48.069 / max:    62.831) us     GPU-0:    52.903 us   +/-  1.017 (min:    51.200 / max:    67.552) us

In [2]: import cupy as cp
   ...: from cupyx.profiler import benchmark
   ...: 
   ...: def f(a, b):
   ...:     return 3 * cp.sin(-a) * b
   ...: 
   ...: for n in [10, 100, 1000, 1000]:
   ...:    a = 0.5 - cp.random.random((n,))
   ...:    b = cp.random.random((n,))
   ...:    print(benchmark(f, (a, b), n_repeat=1000))
   ...: 
f                   :    CPU:    66.001 us   +/- 15.602 (min:    38.402 / max:   115.901) us     GPU-0:    71.266 us   +/- 16.489 (min:    41.440 / max:   122.464) us
f                   :    CPU:    40.369 us   +/-  0.802 (min:    39.240 / max:    53.020) us     GPU-0:    44.104 us   +/-  1.001 (min:    42.656 / max:    56.800) us
f                   :    CPU:    39.809 us   +/-  0.745 (min:    38.641 / max:    46.888) us     GPU-0:    43.662 us   +/-  0.935 (min:    41.728 / max:    51.200) us
f                   :    CPU:    40.016 us   +/-  0.899 (min:    38.831 / max:    55.661) us     GPU-0:    43.793 us   +/-  1.090 (min:    41.696 / max:    59.680) us
```

asi1024 commented:
/test mini

ev-br commented:
Thanks for the review @asi1024 ! I believe I addressed all your comments + added several small fixes for tests which were failing on CI, and a skip for what appears to be a numpy 1.26 bug, fixed in 2.0.
Remaining CI failures are all on numpy < 1.24 which does not have `_set_promotion_state`. So maybe it's about time to bump not only the minimum scipy  version (https://github.com/cupy/cupy/pull/8323#issuecomment-2270502040), but also a minimum numpy version to 1.25 or thereabouts.

asi1024 commented on tests/cupy_tests/core_tests/test_elementwise.py:
```diff
@@ -92,14 +92,14 @@ class TestElementwiseType(unittest.TestCase):
     def test_large_int_upper_1(self, xp, dtype):
         a = xp.array([0], dtype=xp.int8)
-        b = xp.iinfo(dtype).max
+        b = xp.asarray(xp.iinfo(dtype).max)
```
Sorry, I overlooked this during my previous review.
These tests are for the type promotion of ndarray and Python primitive scalar. could you please fix it as follows to check if numpy and cupy throw the same exception?

```py
    @testing.for_int_dtypes(no_bool=True)
    @testing.numpy_cupy_array_equal(accept_error=OverflowError)
    def test_large_int_upper_1(self, xp, dtype):
        a = xp.array([0], dtype=xp.int8)
        b = xp.iinfo(dtype).max
        return a + b
```

ev-br replied:
OK, done.

asi1024 replied:
Could you fix also test_large_int_upper_{2,3,4}?

ev-br replied:
OK, fixed test_large_int_upper_2 and test_large_int_lower_{1,2}. 
_{3,4} variants are identical here and in main and all pass, at least locally.

asi1024 commented:
We are planning to merge a pull request that drops support for NumPy 1.22 and 1.23 in few days. https://github.com/cupy/cupy/pull/8491

asi1024 commented:
/test full

asi1024 commented:
@ev-br In numpy==1.24, the TestFusionScalar::test_numpy_scalar_l test case causes a SEGV during fusion trace due to a bug in NumPy.

Reproducer:

```py
import numpy as np
import cupy as cp

np._set_promotion_state("weak")

@cp.fuse
def func(array):
    np_scalar = np.complex128(1)
    return np_scalar + array

array = cp.arange(12, dtype=np.complex128).reshape(3, 4)
print(func(array))
```

asi1024 commented:
Could you add `@testing.with_requires('numpy>=1.25')` to `test_numpy_scalar_l`?

asi1024 commented on tests/cupy_tests/linalg_tests/test_product.py:
```diff
@@ -468,9 +469,12 @@ def test_matrix_power_large(self, xp, dtype):
         a = xp.eye(23, k=17, dtype=dtype) + xp.eye(23, k=-6, dtype=dtype)
         return xp.linalg.matrix_power(a, 123456789123456789)
 
+    @pytest.mark.skipif(sys.platform == "nt",
```
```suggestion
    @pytest.mark.skipif(sys.platform == "win32",
```

ev-br commented:
Done @asi1024

asi1024 commented:
/test full

asi1024 commented:
These errors seem to be dtype mismatches that occur only with NumPy 1.24. It appears acceptable to skip them if `numpy<1.25`.

```
=========================== short test summary info ============================
FAILED cupy_tests/functional_tests/test_vectorize.py::TestVectorizeExprs::test_vectorize_ufunc_call_dtype
FAILED cupy_tests/lib_tests/test_polynomial.py::TestPolyvalDtypesCombination::test_polyval_diff_types_array_scalar
= 2 failed, 104003 passed, 6627 skipped, 1503 deselected, 718 xfailed, 2559 warnings in 3773.62s (1:02:53) =
```

ev-br commented:
thanks @asi1024 , the last commit adds these skips.

asi1024 commented:
@ev-br Pushed two commits that add skips and relax tolerances.

asi1024 commented:
/test full

asi1024 reviewed (approved):
@ev-br LGTM! Thank you for your contribution!

ev-br commented:
Thank you @asi1024 !
